tmail-backend/mailbox/encrypted/memory/src/main/java/com/linagora/tmail/encrypted/InMemoryEncryptedEmailContentStore.scala

```diff
@@ -30,54 +32,47 @@ class InMemoryEncryptedEmailContentStore @Inject()(blobStore: BlobStore,
-      .getOrElse(throw MessageNotFoundException(messageId))
+      .getOrElse(throw MessageNotFoundException(messageId)))
 
-  override def retrieveDetailedView(messageId: MessageId): EncryptedEmailDetailedView =
-    emailContentStore.get(messageId)
+  override def retrieveDetailedView(messageId: MessageId): Publisher[EncryptedEmailDetailedView] =
+    SMono.just(emailContentStore.get(messageId)
       .map(encryptedEmailContent => EncryptedEmailDetailedView.from(encryptedEmailContent))
-      .getOrElse(throw MessageNotFoundException(messageId))
+      .getOrElse(throw MessageNotFoundException(messageId)))
 
-  override def retrieveAttachmentContent(messageId: MessageId, position: Int): BlobId = {
-    messageIdBlobIdStore.get(messageId)
+  override def retrieveAttachmentContent(messageId: MessageId, position: Int): Publisher[BlobId] =
+    SMono.just(messageIdBlobIdStore.get(messageId)
       .flatMap(positionBlobIdMap => positionBlobIdMap.get(position))
-      .getOrElse(throw AttachmentNotFoundException(messageId, position))
-  }
+      .getOrElse(throw AttachmentNotFoundException(messageId, position)))
 
-  private def storeAttachment(messageId: MessageId, encryptedAttachmentContents: List[String]): Unit = {
+  private def storeAttachment(messageId: MessageId, encryptedAttachmentContents: List[String]): SMono[Unit] = {
     Preconditions.checkNotNull(encryptedAttachmentContents)
-    if (encryptedAttachmentContents.nonEmpty) {
-      val position: AtomicInteger = new AtomicInteger(POSITION_NUMBER_START_AT)
-      val positionBlobIdMap: Map[Int, BlobId] = encryptedAttachmentContents.map(attachment => {
-        val blobId: BlobId =
-          SMono.fromPublisher(blobStore.save(bucketName, attachment.getBytes(StandardCharsets.UTF_8), storagePolicy))
-            .block()
-        position.getAndIncrement() -> blobId
-      }).toMap
-      messageIdBlobIdStore.put(messageId, positionBlobIdMap)
-    }
+    val position: AtomicInteger = new AtomicInteger(POSITION_NUMBER_START_AT)
+    SFlux.fromIterable(encryptedAttachmentContents)
+      .concatMap(attachmentContent => SMono.fromPublisher(blobStore.save(bucketName, attachmentContent.getBytes(StandardCharsets.UTF_8), storagePolicy)))
+      .map(blobId=> position.getAndIncrement()-> blobId)
+      .collectSeq()
+      .map(positionBlobIdList=> positionBlobIdList.toMap)
+      .map(positionBlobIdMap => messageIdBlobIdStore.put(messageId, positionBlobIdMap))
+      .`then`()
   }
 
-  private def removeAttachment(messageId: MessageId): Unit = {
-    messageIdBlobIdStore.get(messageId) match {
-      case Some(positionBlobIdMap) => positionBlobIdMap.values
-        .toList
-        .foreach(blobId => SMono.fromPublisher(blobStore.delete(bucketName, blobId))
-          .block())
-      case _ =>
-    }
-    messageIdBlobIdStore.remove(messageId)
-  }
+  private def removeAttachment(messageId: MessageId): SMono[Unit] =
+    deleteBlobStore(messageId)
+      .collectSeq()
```

Comment:
You don't need the collectSeq if you don't use it. I think the `then` is enough to switch to SMono